confluent_kafka/consumer.py

```diff
@@ -0,0 +1,162 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+#
+# Copyright 2018 Confluent Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from .cimpl import Consumer as _impl
+from warnings import warn
+
+
+class Consumer(_impl):
+    """
+        Create a new Kafka Consumer instance.
+
+        To avoid spontaneous calls from non-Python threads all callbacks will only be served upon
+        calling ```client.poll()``` or ```client.flush()```.
+
+        :param dict conf: Configuration properties.
+            See https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md for more information.
+        :param func key_deserializer(topic, key): Converts message key bytes to object.
+            **note** deserializers are responsible for handling NULL keys
+        :param func value_deserializer(topic, value): Converts message value bytes to object.
+            **note** deserializers are responsible for handling NULL values
+        :param func on_commit(err, [partitions]): Callback used to indicate success or failure
+            of an offset commit.
+        :param func stats_cb(json_str): Callback for statistics emitted every ``statistics.interval.ms``.
+            See https://github.com/edenhill/librdkafka/wiki/Statistics” for more information.
+        :param func throttle_cb(confluent_kafka.ThrottleEvent): Callback for throttled request reporting.
+        :param logging.handlers logger: Forwards logs from the Kafka client to the provided handler instance.
+            Log messages will only be forwarded when ``client.poll()`` or ``producer.flush()`` are called.
+        :raises TypeError: If conf is not a dict.
+    """
+    def __new__(cls, *args, **kwargs):
```

Comment:
fancy. this seems like a good solution to the requirements to me, though the scenario is unique to python and i'm not 100% confident i'm not missing any downsides.

Reply:
It's not terribly easy to read but it does allow me to offer both APIs at almost no cost to those who wish to forego using the serializing producer/consumer

Reply:
Here's what the manual says on `__new__`:

>__new__() is intended mainly to allow subclasses of immutable types (like int, str, or tuple) to customize instance creation. It is also commonly overridden in custom metaclasses in order to customize class creation.

this is neither of those use cases, which makes me wary. I still **think** it's justified as a performance optimization (but again, wary that there might be something important i don't know about). i'd be interested in the perf difference between using a byte deserializer and bypassing the `DeserializingConsumer` as you're doing now as the former is more conservative.

Reply:
>It is also commonly overridden in custom metaclasses in order to customize class creation.

Technically the extension's implementation is not a metaclass no but is this not an example of customizing class creation?